common/components/GenerateKeystoreModal/index.tsx

```diff
@@ -0,0 +1,160 @@
+    keystoreFile: null,
+    hasError: false
+  };
+
+  constructor(props: Props) {
+    super(props);
+
+    if (props.privateKey) {
+      this.state = {
+        ...this.state,
+        privateKey: props.privateKey
+      };
+    }
+  }
+
+  public render() {
+    const {
+      privateKey,
+      password,
+      isPrivateKeyVisible,
+      isPasswordVisible,
+      keystoreFile,
+      hasError
+    } = this.state;
+
+    const isPrivateKeyValid = isValidPrivKey(privateKey);
+    const isPasswordValid = password.length >= MINIMUM_PASSWORD_LENGTH;
+
+    return (
+      <Modal
+        title={translate('Generate Keystore File')}
+        isOpen={true}
+        handleClose={this.props.handleClose}
+      >
+        <form className="GenKeystore" onSubmit={this.handleSubmit}>
+          <label className="GenKeystore-field">
+            <h4 className="GenKeystore-field-label">Private Key</h4>
+            <Input
+              isValid={isPrivateKeyValid}
+              isVisible={isPrivateKeyVisible}
+              name="privateKey"
+              value={privateKey}
+              handleInput={this.handleInput}
+              placeholder="f1d0e0789c6d40f39..."
+              handleToggle={this.togglePrivateKey}
+              disabled={!!this.props.privateKey}
+            />
+          </label>
+          <label className="GenKeystore-field">
+            <h4 className="GenKeystore-field-label">Password</h4>
+            <Input
+              isValid={isPasswordValid}
+              isVisible={isPasswordVisible}
+              name="password"
+              value={password}
+              placeholder={translateRaw('Minimum 9 characters')}
+              handleInput={this.handleInput}
+              handleToggle={this.togglePassword}
+            />
+          </label>
+
+          {!keystoreFile ? (
+            <button
+              className="GenKeystore-button btn btn-primary btn-block"
+              disabled={!isPrivateKeyValid || !isPasswordValid}
+            >
+              {translate('Generate Keystore File')}
+            </button>
+          ) : hasError ? (
+            <p className="alert alert-danger">
+              Keystore generation failed or was invalid. In order to prevent loss of funds, we
+              cannot provide you with a keystore file that may be corrupted. Refresh the page or use
+              a different browser, and try again.
+            </p>
+          ) : (
+            <a
+              onClick={this.props.handleClose}
+              href={keystoreFile.blob}
+              className="GenKeystore-button btn btn-success btn-block"
+              aria-label={translateRaw('x_Keystore')}
+              aria-describedby={translateRaw('x_KeystoreDesc')}
+              download={keystoreFile.filename}
+            >
+              {translate('Download Keystore File')}
+            </a>
+          )}
+        </form>
+      </Modal>
+    );
+  }
+
+  private togglePrivateKey = () => {
```

Comment:
Minor thing for `togglePrivateKey` and `togglePassword`, you may want to use a callback function instead of setting directly from current state to guarantee that the proper previous state is gotten.

Reply:
The setState callback is only used for when you need to call something that subsequently uses state. I.e.

```
state = {
  isTrue: false,
};

someFunction() {
  this.setState({ isTrue: true });
  someOtherFunction();
}

someOtherFunction() {
  if (this.state.isTrue) {
    doSomething();
  }
}
```
In this case, you'd want `someOtherFunction to run only after this.state.isTrue has changed. In my case, there's no chance that `showPrivateKey` or whatever will have changed, since this is the only function that changes it.

Reply:
Really? I thought if you trigger the same `setState` function twice really quickly theres a chance they'll be batched together, both using the same old state. Then you wouldn't get the behaviour you wanted.

Reply:
I believe Henry is correct - I remember reading an article about this.

Here it is: https://medium.com/javascript-scene/setstate-gate-abc10a9b2d82

Reply:
I still think this behavior is desired. If it's batching it, and hasn't set the new state yet, then the UI still shows the private key as hidden. So assuming a user were able to click twice faster than the state changed, and the component re-rendered (_that_ happens synchronously) both clicks would have been on a "reveal private key" button.

Though this would definitely _not_ be desired if, say, one thing called `toggle` to make it true, and then another piece of logic called `toggle` again to turn it back to false. But in those cases, I tend to not use `toggle` and instead have one for making it true and another for making it false.

Reply:
Good point - batching in this case wouldn't actually do any harm.